crates/walrus-service/src/client/daemon/routes.rs

```diff
@@ -384,6 +397,149 @@ pub(super) async fn get_blob_by_object_id<T: WalrusReadClient>(
+    #[rest_api_error(reason = "INVALID_BYTE_RANGE", status = ApiStatusCode::InvalidArgument)]
+    InvalidByteRange { message: String },
+
+    /// The blob size exceeds the maximum allowed size that was configured for this service.
+    #[error("the blob size exceeds the maximum allowed size: {0}")]
+    #[rest_api_error(reason = "BLOB_TOO_LARGE", status = ApiStatusCode::SizeExceeded)]
+    BlobTooLarge(u64),
+
+    #[error(transparent)]
+    #[rest_api_error(delegate)]
+    Internal(#[from] anyhow::Error),
+}
+
+impl From<ClientError> for ByteRangeReadError {
+    fn from(error: ClientError) -> Self {
+        match error.kind() {
+            ClientErrorKind::BlobIdDoesNotExist => Self::BlobNotFound,
+            ClientErrorKind::BlobIdBlocked(_) => Self::Blocked,
+            ClientErrorKind::BlobTooLarge(max_blob_size) => Self::BlobTooLarge(*max_blob_size),
+            ClientErrorKind::ByteRangeReadError(msg) => Self::InvalidByteRange {
+                message: msg.to_string(),
+            },
+            _ => anyhow::anyhow!(error).into(),
+        }
+    }
+}
+
+/// Retrieve a specific byte range from a Walrus blob.
+///
+/// Reads a specific portion of a blob identified by the blob ID, starting from a given byte
+/// position and reading a specified number of bytes. This is more efficient than retrieving the
+/// entire blob when only a portion is needed.
+///
+/// # Query Parameters
+/// - `start`: The starting byte position (0-indexed, inclusive)
+/// - `length`: The number of bytes to read
+///
+/// # Validation
+/// - `start` and `length` must not overflow when added together
+/// - The byte range `[start, start + length)` must be within the blob's size
+/// - `length` must be greater than 0
+///
+/// # Example
+/// ```bash
+/// # Read 1024 bytes starting from position 0
+/// curl "$AGGREGATOR/v1/blobs/{blob_id}/byte-range?start=0&length=1024"
+/// ```
+#[tracing::instrument(
+    level = Level::ERROR,
+    skip_all,
+    fields(
+        %blob_id,
+        start=%params.start,
+        length=%params.length,
+    ),
+)]
+#[utoipa::path(
+    get,
+    path = BLOB_BYTE_RANGE_GET_ENDPOINT,
+    params(
+        ("blob_id" = BlobId,),
+        ByteRangeQueryParams,
+    ),
+    responses(
+        (status = 200, description = "The byte range was read successfully", body = [u8]),
+        ByteRangeReadError,
+    ),
+)]
+pub(super) async fn get_blob_byte_range<T: WalrusReadClient>(
+    request_method: Method,
+    request_headers: HeaderMap,
+    State(client): State<Arc<T>>,
+    Path(BlobIdString(blob_id)): Path<BlobIdString>,
+    Query(params): Query<ByteRangeQueryParams>,
+) -> Response {
+    tracing::debug!(
+        "starting to read byte range from blob: start={}, length={}",
+        params.start,
+        params.length
+    );
+
+    // Validate parameters upfront
+    if params.length == 0 {
+        return ByteRangeReadError::InvalidByteRange {
+            message: "length must be greater than 0".to_string(),
```

Comment:
Also mozilla's website says out of bound response should get 416 https://developer.mozilla.org/en-US/docs/Web/HTTP/Guides/Range_requests